ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Client.hs

```diff
@@ -0,0 +1,128 @@
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.Protocol.TxSubmission.Client
+  ( TxSubmissionClientPipelined (..)
+  , TxSubmissionSender (..)
+  , txSubmissionClientPeerPipelined
+  , txSubmissionSender
+  )
+  where
+
+import           Numeric.Natural (Natural)
+
+import           Network.TypedProtocol.Core
+import           Network.TypedProtocol.Pipelined
+
+import           Ouroboros.Network.Protocol.TxSubmission.Type
+
+--
+-- Pipelined client
+--
+
+data TxSubmissionClientPipelined hash tx m a where
+  TxSubmissionClientPipelined
+    :: TxSubmissionSender hash tx Z c m a
+    -> TxSubmissionClientPipelined hash tx m a
+
+-- |
+-- @'TxSubmissionSender'@ can be transformed into a pipelined @'PeerSender'@.
+-- We are only pipelining trasactions requests, e.g. @'MsgTx'@.
+--
+data TxSubmissionSender hash tx (n :: N) c m a where
+
+  -- |
+  -- Ask the server for a list of transaction hashes by sending @'MsgGetHashes'@.
+  --
+  SendMsgGetHashes
+    :: Natural
+    -> ([hash] -> m (TxSubmissionReceiver hash tx Z c m a))
+    -> TxSubmissionSender hash tx Z c m a
+
+  -- |
+  -- Terminate the tx-submission protocol.
+  --
+  SendMsgMsgDone
+    :: a
+    -> TxSubmissionSender hash tx Z c m a
+
+-- |
+-- Transaction receiver. @c@ can be an accumulator (e.g. a monoid) which is
+-- updated for every incoming transaction.
+--
+data TxSubmissionReceiver hash tx (n :: N) c m a where
+
+  -- |
+  -- Pipeline @'MsgGetTx'@.
+  --
+  SendMsgGetTxPipelined
+    :: hash
+    -> c
+    -> (Maybe tx -> c -> m c)
+    -> TxSubmissionReceiver hash tx (S n) c m a
+    -> TxSubmissionReceiver hash tx    n  c m a
+
+  -- |
+  -- Collect transactions that arrived as a response to @'MsgGetTx'@.
+  -- 
+  CollectPipelined
+    :: Maybe (TxSubmissionReceiver hash tx (S n) c m a)
+    -> (c ->  TxSubmissionReceiver hash tx    n  c m a)
+    ->        TxSubmissionReceiver hash tx (S n) c m a
+
+
+-- |
+-- Transform a @'TxSubmissionClientPipelined'@ into a @'PeerPipelined'@ which
+-- pipelines @'MsgGetTx'@ messages.
+--
+txSubmissionClientPeerPipelined
+    :: forall hash tx m a.
+       Functor m
+    => TxSubmissionClientPipelined hash tx m a
+    -> PeerPipelined (TxSubmission hash tx) AsClient StIdle m a
+txSubmissionClientPeerPipelined (TxSubmissionClientPipelined sender) =
+  PeerPipelined $ txSubmissionSender sender
+
+
+-- |
+-- The @'PeerSender'@ which asks for available transaction hashes and pipelines
+-- @'MsgGetTx'@s.
+--
+txSubmissionSender
+    :: forall hash tx (n :: N) c m a.
+       Functor m
+    => TxSubmissionSender hash tx n c m a
+    -> PeerSender (TxSubmission hash tx) AsClient StIdle n c m a
+
+txSubmissionSender (SendMsgMsgDone a) =
+    SenderYield (ClientAgency TokIdle) MsgDone (SenderDone TokDone a)
+
+txSubmissionSender (SendMsgGetHashes n receiver) =
+    SenderYield
+      (ClientAgency TokIdle)
+      (MsgGetHashes n)
+      (SenderAwait (ServerAgency TokSendHashes)
```

Comment:
Can we do this without adding `SenderAwait`? I suspect we can. Is it that much more horrible?

Reply:
I pushed a version with it.  Since we anyway want to pipeline both `MsgGetHashes` and `MsgGetTx` this does not buy us much (and it would just eliminate a single `error "impossible happend"`).  But on the other side, we could end up with a single `Peer` in `Core` rather than both `Peer` and `SenderPeer`.